crates/oxc_query/src/vertex.rs

```diff
@@ -0,0 +1,319 @@
+use std::rc::Rc;
+
+use enum_as_inner::EnumAsInner;
+use oxc_ast::{ast::*, AstKind};
+use oxc_semantic::{AstNode, AstNodeId};
+use oxc_span::{GetSpan, Span};
+use url::Url;
+
+use crate::util::{expr_to_maybe_const_string, jsx_attribute_to_constant_string};
+
+#[derive(Debug, Clone, EnumAsInner)]
+pub enum Vertex<'a> {
+    ASTNode(AstNode<'a>),
+    AssignmentType(&'a BindingPatternKind<'a>),
+    Class(Rc<ClassVertex<'a>>),
+    ClassMethod(Rc<ClassMethodVertex<'a>>),
+    ClassProperty(Rc<ClassPropertyVertex<'a>>),
+    DefaultImport(&'a ImportDefaultSpecifier),
+    Expression(&'a Expression<'a>),
+    File,
+    Import(Rc<ImportVertex<'a>>),
+    Interface(Rc<InterfaceVertex<'a>>),
+    InterfaceExtend(Rc<InterfaceExtendVertex<'a>>),
+    JSXAttribute(&'a JSXAttribute<'a>),
+    JSXElement(Rc<JSXElementVertex<'a>>),
+    JSXExpressionContainer(&'a JSXExpressionContainer<'a>),
+    JSXFragment(&'a JSXFragment<'a>),
+    JSXOpeningElement(&'a JSXOpeningElement<'a>),
+    JSXSpreadAttribute(&'a JSXSpreadAttribute<'a>),
+    JSXSpreadChild(&'a JSXSpreadChild<'a>),
+    JSXText(&'a JSXText),
+    ObjectLiteral(&'a ObjectExpression<'a>),
+    PathPart(usize),
+    SearchParameter(Rc<SearchParameterVertex>),
+    Span(Span),
+    SpecificImport(&'a ImportSpecifier),
+    TypeAnnotation(Rc<TypeAnnotationVertex<'a>>),
+    Type(&'a TSType<'a>),
+    URL(Rc<Url>),
+    VariableDeclaration(Rc<VariableDeclarationVertex<'a>>),
+    ReturnStatementAST(Rc<ReturnStatementVertex<'a>>),
+}
+
+impl<'a> Vertex<'a> {
+    pub fn make_type_name(&self) -> &'static str {
+        match self {
+            Vertex::ASTNode(_) => "ASTNode",
+            Vertex::AssignmentType(_) => "AssignmentType",
+            Vertex::Class(class) => class.type_name(),
+            Vertex::ClassMethod(_) => "ClassMethod",
+            Vertex::ClassProperty(_) => "ClassProperty",
+            Vertex::DefaultImport(_) => "DefaultImport",
+            Vertex::Expression(_) => "Expression",
+            Vertex::File => "File",
+            Vertex::Import(import) => import.type_name(),
+            Vertex::Interface(iface) => iface.type_name(),
+            Vertex::InterfaceExtend(iex) => match **iex {
+                InterfaceExtendVertex::Identifier(_) => "SimpleExtend",
+                InterfaceExtendVertex::MemberExpression(_) => "MemberExtend",
+            },
```

Comment:
```suggestion
            Vertex::InterfaceExtend(InterfaceExtendVertex::Identifier(..)) => "SimpleExtend",
            Vertex::InterfaceExtend(InterfaceExtendVertex::MemberExpression(..)) => "MemberExtend",
```
pattern matching works in nested fashion too

Reply:
I can't do this because the iex is an Rc

Reply:
Ah that's unfortunate.